iwyu.cc

```diff
@@ -3732,11 +3732,20 @@ class IwyuAstConsumer
   //    template <class T> struct Foo;
   //    template<> struct Foo<int> { ... };
   // we don't want iwyu to recommend removing the 'forward declare' of Foo.
+  //
+  // Additionally, this type of decl is also used to represent template
```

Comment:
I remember reading the Josuttis/Vandevoorde book on C++ templates about a decade ago, and they consistently used "specialization" instead of "instantiation", that is, they were the same thing in the standard text back then.

I think "explicit instantiation" is the primary context in which "instantiation" is used, even now. So maybe this comment should say "... is also used to represent *explicit* template instantiations...". Everything else is a specialization, not an instantiation.

Also, the "Additionally ... also" pair looks like a double-take. I think you can remove "also".

Reply:
Yes, I meant to write *explicit* instantiations. Thanks!

Regarding the usage of the terms, I must admit I find the whole thing quite confusing:
Looking at how the [C++ standard](from http://eel.is/c++draft/temp.spec) uses them it seems to me that *instantiation* is the process of substitution and generation, and also the resulting types/methods (ClassTemplateSpecializationDecl in clang), whereas *specializations* are the template definitions before the process of instantiation (the ClassTemplateDecl in clang)

But now looking at clang, `ClassTemplateSpecializationDecl` *is* an instantiated `CXXRecordDecl` whereas `ClassTemplateDecl` *references* the uninstantiated `CXXRecordDecl`, which seems a bit backward to me :)